### PR TITLE
Henryh/incorporate all missing timestep samples

### PR DESCRIPTION
**Pull request by Hgherzog**
This pull request introduces significant updates to the `helios/data/dataset.py` file, focusing on improving the handling of missing data, enhancing flexibility for temporal data, and simplifying the codebase by removing unused methods and imports. The changes aim to make the dataset processing more robust and configurable.

View of Runs: https://wandb.ai/eai-ai2/helios-debug?nw=09v3jf3bj7zo 
New datasets:

presto: `/weka/dfive-default/helios/dataset/presto/h5py_data_w_missing_timesteps_gzip_3/landsat_naip_10_openstreetmap_raster_sentinel1_sentinel2_l2a_srtm_worldcover/117473`

osm_sampling: `/weka/dfive-default/helios/dataset/osm_sampling/h5py_data_w_missing_timesteps_gzip_3/landsat_naip_10_openstreetmap_raster_sentinel1_sentinel2_l2a_srtm_worldcover/285288` 

### Enhancements to Missing Data Handling:
* Introduced `_fill_missing_timesteps` and `_fill_missing_modality` methods to handle missing timesteps and modalities more effectively. These methods allow filling missing data with predefined values or treating modalities with missing timesteps as entirely absent, based on a new configuration flag.
* Added `use_modalities_with_missing_timesteps` and `max_sequence_length` parameters to the dataset class, enabling more control over how missing timesteps are handled and ensuring consistent padding of time dimensions. [[1]](diffhunk://#diff-60cfa8bc3b47a7f0ebed7f1aa5afd2d06e83716a30c9b85ea1a04d170d6ee601L371-L373) [[2]](diffhunk://#diff-60cfa8bc3b47a7f0ebed7f1aa5afd2d06e83716a30c9b85ea1a04d170d6ee601L409-R419)

### Improvements to Temporal Data Handling:
* Implemented `_pad_timestamps` to ensure timestamps are padded to the `max_sequence_length`, maintaining consistency across samples.
* Enhanced the `read_h5_file` method to return both sample data and missing timestep masks, improving support for multi-temporal data. [[1]](diffhunk://#diff-60cfa8bc3b47a7f0ebed7f1aa5afd2d06e83716a30c9b85ea1a04d170d6ee601L708-R702) [[2]](diffhunk://#diff-60cfa8bc3b47a7f0ebed7f1aa5afd2d06e83716a30c9b85ea1a04d170d6ee601L730-R748)

### Codebase Simplification:
* Removed unused methods such as `_log_modality_distribution` and `get_sample_data_for_histogram`, as well as redundant imports like `random` and `tqdm`. This streamlining reduces clutter and improves maintainability. [[1]](diffhunk://#diff-60cfa8bc3b47a7f0ebed7f1aa5afd2d06e83716a30c9b85ea1a04d170d6ee601L541-L590) [[2]](diffhunk://#diff-60cfa8bc3b47a7f0ebed7f1aa5afd2d06e83716a30c9b85ea1a04d170d6ee601L602-L647) [[3]](diffhunk://#diff-60cfa8bc3b47a7f0ebed7f1aa5afd2d06e83716a30c9b85ea1a04d170d6ee601L5) [[4]](diffhunk://#diff-60cfa8bc3b47a7f0ebed7f1aa5afd2d06e83716a30c9b85ea1a04d170d6ee601L17-L33)
* Simplified modality-specific logic in methods such as `fingerprint` and `_filter_sample_indices_for_training` to ensure cleaner and more efficient code. [[1]](diffhunk://#diff-60cfa8bc3b47a7f0ebed7f1aa5afd2d06e83716a30c9b85ea1a04d170d6ee601R453-L459) [[2]](diffhunk://#diff-60cfa8bc3b47a7f0ebed7f1aa5afd2d06e83716a30c9b85ea1a04d170d6ee601L494-R506)

### Refactoring of Spatial Modality Handling:
* Replaced hardcoded modality lists with dynamic checks using `Modality.get` and `is_spatial`, improving the flexibility of the `height` and `width` property methods. [[1]](diffhunk://#diff-60cfa8bc3b47a7f0ebed7f1aa5afd2d06e83716a30c9b85ea1a04d170d6ee601L182-R183) [[2]](diffhunk://#diff-60cfa8bc3b47a7f0ebed7f1aa5afd2d06e83716a30c9b85ea1a04d170d6ee601L203-R203)

### Additional Improvements:
* Updated method signatures and docstrings to reflect new functionality, ensuring better documentation and clarity for future developers. [[1]](diffhunk://#diff-60cfa8bc3b47a7f0ebed7f1aa5afd2d06e83716a30c9b85ea1a04d170d6ee601L371-L373) [[2]](diffhunk://#diff-60cfa8bc3b47a7f0ebed7f1aa5afd2d06e83716a30c9b85ea1a04d170d6ee601R389-L392)

### CODE REVIEW

File: helios/data/dataset.py [hunk 1]
```diff
@@ -538,56 +547,11 @@ def prepare(self) -> None:
...
-        logger.info("H5 files already exist, skipping creation")
-        logger.info(f"H5 files exist in {self.h5py_dir}")
         num_samples = int(self.h5py_dir.name)
         self.latlon_distribution = self.get_geographic_distribution()
         self.sample_indices = np.arange(num_samples)
         self._filter_sample_indices_for_training()
 
-    # TODO: Needs to be gotten or owned from th other class
-
-    def save_latlon_distribution(self, latlons: np.ndarray) -> None:
```
Comment (Hgherzog) [inline helios/data/dataset.py:553]: This is done in h5 conversion now

File: scripts/2025_04_23_benchmarking_ladder/base_galileo_max.py
Comment (Hgherzog) [inline scripts/2025_04_23_benchmarking_ladder/base_galileo_max.py:1]: I should make a seperate file for testing these things rather than just changing this one

File: helios/data/dataset.py [hunk 2]
```diff
@@ -750,35 +763,37 @@ def __getitem__(self, args: GetItemArgs) -> tuple[int, HeliosSample]:
...
+                logger.info(f"Normalizing {modality_name}")
+                modality_data = sample_dict[modality_name]
+                missing_mask = modality_data == MISSING_VALUE
+                normalized_data = self.normalize_image(
+                    Modality.get(modality_name), modality_data
                 )
+                # Sentinel Values must be reset after normalization so they can be recognized by missing mask
+                sample_dict[modality_name] = np.where(
+                    missing_mask, modality_data, normalized_data
+                ).astype(self.dtype)
```
Comment (yawenzzzz) [inline helios/data/dataset.py:797]: I tried last time changing DType to `bfloat16` and it doesn't work well with numpy array. If you prefer using DType and I think it will be nice to train in `bfloat16`, then better to convert numpy array to torch tensor for compatibility with DType.

File: helios/data/dataset.py [hunk 3]
```diff
@@ -663,24 +580,100 @@ def normalize_image(self, modality: ModalitySpec, image: np.ndarray) -> np.ndarr
...
             return self.normalizer_predefined.normalize(modality, image)
 
+    def _fill_missing_timesteps(
+        self,
+        modality_data: np.ndarray,
+        missing_timestep_mask: np.ndarray,
+    ) -> np.ndarray:
+        """Fill the missing timesteps with the missing value."""
+        # cast to appropriate dtype to prevent overflow from missing values
+        modality_data = modality_data.astype(self.dtype)
```
Comment (yawenzzzz) [inline helios/data/dataset.py:590]: Same thing here, DType may not work well with np.ndarray

File: helios/dataset/sample.py [hunk 4]
```diff
@@ -61,29 +61,23 @@ def get_latlon(self) -> np.ndarray:
         lon, lat = transformer.transform(x, y)
         return np.array([lat, lon])
 
-    def get_timestamps(self) -> np.ndarray:
+    def get_timestamps(self) -> dict[ModalitySpec, np.ndarray]:
         """Get the timestamps of the sample."""
         # Assume that all multitemporal modalities have the same timestamps
```
Comment (yawenzzzz) [inline helios/dataset/sample.py]: This is not true anymore given some modality may have missing timesteps.

File: helios/train/train_module/galileo.py [hunk 5]
```diff
@@ -229,6 +229,8 @@ def train_batch(
         self.model.train()
 
         # Set the maximum number of tokens
+        total_space_time_mask_loss = torch.tensor(0.0, device=self.device)
+        total_random_mask_loss = torch.tensor(0.0, device=self.device)
```
Comment (yawenzzzz) [inline helios/train/train_module/galileo.py]: We can apply different masking strategies to a and b, so would be better to use `total_mask_loss_a` and `total_mask_loss_b` to differentiate them, then even if we apply random & random or any other masking strategy, it still applies.

File: helios/train/train_module/galileo.py [hunk 6]
```diff
@@ -280,35 +288,81 @@ def train_batch(
...
+            total_space_time_mask_loss,
+            ReduceType.mean,
+            namespace="train",
+        )
+        self.trainer.record_metric(
+            "random_mask_loss",
+            total_random_mask_loss,
+            ReduceType.mean,
+            namespace="train",
+        )
```
Comment (yawenzzzz) [inline helios/train/train_module/galileo.py]: Save comment as above.

File: scripts/base_debug_scripts/galileo.py [hunk 7]
```diff
@@ -118,20 +122,20 @@ def build_train_module_config(
...
+        Modality.SENTINEL2_L2A.name: int(tiny_model_args["encoder_depth"]),
+        Modality.LATLON.name: int(tiny_model_args["encoder_depth"]),
+        Modality.SENTINEL1.name: int(tiny_model_args["encoder_depth"]),
         Modality.WORLDCOVER.name: 0,
-        Modality.SRTM.name: 2,
+        Modality.SRTM.name: int(tiny_model_args["encoder_depth"] // 2),
         Modality.OPENSTREETMAP_RASTER.name: 0,
-        Modality.LANDSAT.name: 4,
+        Modality.LANDSAT.name: int(tiny_model_args["encoder_depth"]),
     }
```
Comment (yawenzzzz) [inline scripts/base_debug_scripts/galileo.py:132]: Thanks for the fix!

File: helios/data/dataset.py (hunk 2 above)
Comment (Hgherzog) [reply helios/data/dataset.py:797]: reverted
